cmake: Only use freestanding flag on non-posix platform

tejlmand opened the pull request:
Zephyr build system has always added `-ffreestanding` got gcc based
compilers, and then arch/posix/CMakeLists.txt would append
`-fno-freestanding` to Zephyr compile options.

This works well for gcc based toolchains as flags where a later
`-fno-freestanding` would overrule an earlier `-ffreestanding` flag.

With LLVM support, this is no longer possible, as clang doesn't support
a `-fhosted` / `-fno-freestanding` flag, and thus `-ffreestanding` would
always be added as compile flag when using clang for native_posix
builds.

This is now fix by only setting the freestanding compilation flag on
non posix SoCs.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Only use freestanding flag on non-posix platform

Zephyr build system has always added `-ffreestanding` got gcc based
compilers, and then arch/posix/CMakeLists.txt would append
`-fno-freestanding` to Zephyr compile options.

This works well for gcc based toolchains as flags where a later
`-fno-freestanding` would overrule an earlier `-ffreestanding` flag.

With LLVM support, this is no longer possible, as clang doesn't support
a `-fhosted` / `-fno-freestanding` flag, and thus `-ffreestanding` would
always be added as compile flag when using clang for native_posix
builds.

This is now fix by only setting the freestanding compilation flag on
non posix SoCs.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
@nashif this should fix the issue with freestanding being set as compile flag on native posix builds you mentioned at the Toolchain wg meeting.

nashif commented:
@tejlmand  it does not fix the issue:

```
FAILED: zephyr/CMakeFiles/offsets.dir/arch/posix/core/offsets/offsets.c.obj
ccache /usr/lib64/ccache/clang -DBUILD_VERSION=zephyr-v2.4.0-1586-ge27a8878d35c -DKERNEL -D_POSIX_C_SOURCE=200809 -D_XOPEN_SOURCE=60c
clang-10: error: unknown argument: '-fno-freestanding'
```

tejlmand commented:
> 
> 
> @tejlmand it does not fix the issue:
> 
> ```
> FAILED: zephyr/CMakeFiles/offsets.dir/arch/posix/core/offsets/offsets.c.obj
> ccache /usr/lib64/ccache/clang -DBUILD_VERSION=zephyr-v2.4.0-1586-ge27a8878d35c -DKERNEL -D_POSIX_C_SOURCE=200809 -D_XOPEN_SOURCE=60c
> clang-10: error: unknown argument: '-fno-freestanding'
> ```

Oh, then you see a different issue than what I experiences.
I don't see this error.

Let me try to see if I can reproduce this.

tejlmand commented:
> > @tejlmand it does not fix the issue:
> > ```
> > FAILED: zephyr/CMakeFiles/offsets.dir/arch/posix/core/offsets/offsets.c.obj
> > ccache /usr/lib64/ccache/clang -DBUILD_VERSION=zephyr-v2.4.0-1586-ge27a8878d35c -DKERNEL -D_POSIX_C_SOURCE=200809 -D_XOPEN_SOURCE=60c
> > clang-10: error: unknown argument: '-fno-freestanding'
> > ```
> 
> Oh, then you see a different issue than what I experiences.
> I don't see this error.
> 
> Let me try to see if I can reproduce this.

@nashif  Still getting this:
```
[7/83] ccache /opt/llvm/clang+llvm-10.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/clang -DBUILD_VERSION=zephyr-v2.4.0-1585-g83e555088f36 -DKERNEL -D_POSIX_C_SOURCE=200809 -D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED -D__ZEPHYR__=1 -I../kernel/include -I../arch/posix/include -I../include -Izephyr/include/generated -I../soc/posix/inf_clock -I../boards/posix/native_posix -Os -imacros /projects/github/ncs/zephyr/build/zephyr/include/generated/autoconf.h -fno-common -g -imacros /projects/github/ncs/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-typedef-redefinition -Wno-pointer-sign -Wpointer-arith -Wno-sometimes-uninitialized -Wno-shift-overflow -Wno-missing-braces -Wno-self-assign -Wno-address-of-packed-member -Wno-unused-function -Wno-initializer-overrides -Wno-section -Wno-unknown-warning-option -Wno-unused-variable -Wno-format-invalid-specifier -Wno-gnu -Wno-tautological-compare -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fmacro-prefix-map=/projects/github/ncs/zephyr/samples/hello_world=CMAKE_SOURCE_DIR -fmacro-prefix-map=/projects/github/ncs/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/projects/github/ncs=WEST_TOPDIR -ffunction-sections -fdata-sections -m32 -include /projects/github/ncs/zephyr/arch/posix/include/posix_cheats.h -std=c11 -MD -MT zephyr/CMakeFiles/offsets.dir/arch/posix/core/offsets/offsets.c.obj -MF zephyr/CMakeFiles/offsets.dir/arch/posix/core/offsets/offsets.c.obj.d -o zephyr/CMakeFiles/offsets.dir/arch/posix/core/offsets/offsets.c.obj   -c /projects/github/ncs/zephyr/arch/posix/core/offsets/offsets.c
```
which looks correct.

Settings:
```
$ echo $ZEPHYR_TOOLCHAIN_VARIANT
llvm
$ echo $LLVM_TOOLCHAIN_PATH
/opt/llvm/clang+llvm-10.0.0-x86_64-linux-gnu-ubuntu-18.04/
```

which sample and board are you building for ?
On my side:
```
cmake -GNinja -DBOARD=native_posix -Bbuild samples/hello_world/
```

tejlmand committed:
cmake: llvm/clang compiler flag for hosted updated

clang compiler flags now sets the hosted property directly instead of
a check and set.

There should be no need to do a check on an empty property value.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
@nashif I've noticed that clang compiler flag does a `check_set_compiler_property()` on the empty hosted property.
I have changed this into a pure `set_compiler_property()` as there is no flag to check.
This ahas been changed in: https://github.com/zephyrproject-rtos/zephyr/pull/30084/commits/395337680343ceb712465acfa1fa20194b8dd1a9 

If this helps, then I have a suspicion that your toolchain capabilities database might be out-of-date, which may happen if playing too much around when adding extra toolchain support.
You may try to clear: `~/.cache/zephyr/ToolchainCapabilityDatabase/`
and see if the problem vanish on master.

Nevertheless I believe: https://github.com/zephyrproject-rtos/zephyr/pull/30084/commits/395337680343ceb712465acfa1fa20194b8dd1a9 should fix the issue on your side.

daor-oti commented on CMakeLists.txt:
```diff
@@ -204,7 +204,9 @@ endif()
 
 # @Intent: Set compiler specific flag for bare metal freestanding option
-zephyr_compile_options($<TARGET_PROPERTY:compiler,freestanding>)
+if(NOT CONFIG_SOC_POSIX)
```
Could we not find a better way of doing this, than to have very specific tests/cases in the top level CMakeLists.txt file?

tejlmand replied:
I completely agree, this is not something i'm fond of, more a short term solution.

We need somehow we can describe if a build is targeting a hosted or freestanding system, and that would probably be the SoC that should set this.

I'm considering something like:
```
config SOC_BUILD_FREESTANDING
        default y
        help
          Prompt less configuration option for SoCs to define 
          if compilation should be build as hosted (setting=n)
          or freestanding application (setting=y).
```

But I haven't taken a deeper look into the most appropriate long term solution, so therefore the current proposal.

daor-oti replied:
Yes, something like that, but did we not at some point have different toolchain configurations, depending upon building for target or host? We have configurations named target.cmake and generic.cmake, would it not be possible to utilize that, such as to not modify top level CMakeLists.txt with specific configuration flags?

tejlmand replied:
No, the target and generic files are for having just the basic in place for DT.
https://github.com/zephyrproject-rtos/zephyr/commit/a0a63ac3444800de9542fd47e23205ef0ff75468
> Split up the toolchain configuration into two phases, generic and
target. The 'generic' phase configures the toolchain just enough to be
able to preprocess DT files. The 'target' phase completes the
configuration with target-specific configuration.

so this is not related to hosted vs. freestanding application.
Of course we could repurpose those files, but as Zephyr is intended for embedded, i.e. freestanding, then I think it would be too much.

For sure we need to handle building hosted or freestanding in a better way, but especially with llvm, things are not so simple, as it can build both hosted and freestanding using same toolchain. (which gcc also can, but with gcc you have dedicated compiler for cross-compiling --> freestanding).
And on gcc we have the `host` toolchain which simply uses compiler `host-gcc`, but that is not so elegant and I would prefer not to extend that principle into llvm.

I hope you see why I decided a short term proposal that simply uses what is already there, and then look into a more proper solution long term.

Keep posting good suggestions :smile:

daor-oti replied:
> No, the target and generic files are for having just the basic in place for DT.

Yes, I see, I had forgotten that. True, that has no relation to host or target selection, then. It might though, be an idea to make host and target selection similar to the DT case, and add a host.cmake file for property settings which are different from target. It could even take the target.cmake as a basis, and then modify the needed properties.
But, as you state:
>  but as Zephyr is intended for embedded, i.e. freestanding

I do agree that it will probably be too much, and actually complicate the process more, which is exactly what we are trying to avoid.

You are correct, it was the `host-gcc` way/split I was thinking about, but that is probably not a good solution either. 

I agree that your current simple solution for llvm and posix can be a start. I do however think, that we should plan to reconsider that at a later state, so that we do not end up with having many special ifs in the top level CMakeLists.txt file, or even many different kconfigs, which will make it difficult/impossible to add new/other toolchains later on.

With regards to:

> config SOC_BUILD_FREESTANDING

I think that is to specific, and it may be better to simply state the intended target, such as host or target with:
> config SOC_TARGET_HOST

or something along those lines. But we can take that at another time, so we do not stall basic llvm for too long.
